extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java

```diff
@@ -69,44 +63,11 @@ public void handle(RoutingContext event) {
                 //we put the authenticator into the routing context so it can be used by other systems
                 event.put(HttpAuthenticator.class.getName(), authenticator);
 
-                //register the default auth failure handler
-                //if proactive auth is used this is the only one
-                //if using lazy auth this can be modified downstream, to control authentication behaviour
-                event.put(QuarkusHttpUser.AUTH_FAILURE_HANDLER, new BiConsumer<RoutingContext, Throwable>() {
-                    @Override
-                    public void accept(RoutingContext routingContext, Throwable throwable) {
-                        throwable = extractRootCause(throwable);
-                        //auth failed
-                        if (throwable instanceof AuthenticationFailedException) {
-                            authenticator.sendChallenge(event).subscribe().with(new Consumer<Boolean>() {
-                                @Override
-                                public void accept(Boolean aBoolean) {
-                                    if (!event.response().ended()) {
-                                        event.response().end();
-                                    }
-                                }
-                            }, new Consumer<Throwable>() {
-                                @Override
-                                public void accept(Throwable throwable) {
-                                    event.fail(throwable);
-                                }
-                            });
-                        } else if (throwable instanceof AuthenticationCompletionException) {
-                            log.debug("Authentication has failed, returning HTTP status 401");
-                            event.response().setStatusCode(401);
-                            event.response().end();
-                        } else if (throwable instanceof AuthenticationRedirectException) {
-                            AuthenticationRedirectException redirectEx = (AuthenticationRedirectException) throwable;
-                            event.response().setStatusCode(redirectEx.getCode());
-                            event.response().headers().set(HttpHeaders.LOCATION, redirectEx.getRedirectUri());
-                            event.response().headers().set(HttpHeaders.CACHE_CONTROL, "no-store");
-                            event.response().headers().set("Pragma", "no-cache");
-                            event.response().end();
-                        } else {
-                            event.fail(throwable);
-                        }
-                    }
-                });
+                //register the default auth failure handler if no-one provided one up in the stream
```

Comment:
What I'm a bit concerned about here is that it is quite easy now for users to break things, for example, OIDC `AuthenticationFailedException`, if thrown by `quarkus-oidc` is better be handled by `quarkus-oidc` to prepare a correct challenge,  and if the user has registered a mapper then this mapper is in total control now. For ex, a few users combine Basic and OIDC code flow, so if the mapper will blindly return 401 then it is a problem.
I know, when it was not possible to customize with the proactive auth being on, it was a problem, and now that it is fixed, I'm looking for new problems :-).
I'm not sure, it feels like for some cases, we should just not allow intercepting it. If it is only 401 which is returned by default, then sure, why not let users customize it. If it is a redirect to Keycloak that has to be returned, then the custom mapper will not be able to form a redirect URL correctly - as the way it is done it depends on OIDC config, and other factors

Reply:
It makes sense, but then I think you'll have to accept that behavior will differ between enabled and disabled proactive security as you can't stop users from replacing `QuarkusHttpUser.AUTH_FAILURE_HANDLER` in the routing context. Doesn't it also means users are already (potentially) experiencing such problems if they are using disabled proactive sec.? IMHO such exceptions to behavior must be confusing for users (sometimes you can intercept ex. sometimes you can't..)

On level of `quarkus-vertx-http` we can't determine what is `quarkus-oidc` doing and such a feature as _you can only intercept exception if no-one has added headers_ or redirect didn't happen is not user friendly. BTW aren't they partially already able to adjust the behavior if they provide custom auth mechanism? I can drop a portion covering `proactive=true` if you like (as the issue reproducer had proactive disabled), but I can't define cases when we don't want to intercept ex. myself (at least not with required quality).

Reply:
btw even with custom mapping and `proactive=true`, they still can't hack (customize response) the redirect if the response has been sent, so that should be fine.

Reply:
but they can avoid sending a challenge, yeah, didn't think of that.

Reply:
Sure, but with custom HttpAuthenticationMechanism it is quite hard, and requires a real dedication if someone really wants to customize for ex the redirect.

>btw even with custom mapping and proactive=true, they still can't hack (customize response) the redirect if the response has been sent, so that should be fine.

Can you clarify please this point ?

Reply:
> I.e, if registering a custom Resteasy (Reactive) exception mapper can break the OIDC code flow then it is a flexibility which has become too flexible :-), users should be able to customize the response, ex, add some extra headers in case of redirects, if they want to, but if the clients access Quarkus which combines Basic and OIDC code flow, OIDC challenge is initiated, with `AuthenticatedFailedException` thrown and the user's mapper responds with 401 and the client instead of being redirected to Keycloak, then it will become a real problem, it is a user error, but the users should not be able just to get the whole mechanism down so easily, even unintentionally :-)
> 
> Not sure yet how to avoid, but it makes IMHO to think a bit more about it...

I agree, but than it's a breaking change that will reduce current behavior and should be same for proactive sec. disabled/enabled, but at least 2 extensions (undertow and resteasy classic) are not using the default auth failure handler at all (when disabled), so it will take a lot of changes.

Reply:
Please take your time.

Reply:
Thanks, let me think a bit more about it.
I think what might make sense is for the default auth handler to delegate to the custom one so that it can complete the challenge process if the custom one has not provided it (example, the custom exception mapper checks via RoutingContext if it is OIDC which initiated the challenge and if yes then it rethrows the exception ?)

Reply:
> I.e, if registering a custom Resteasy (Reactive) exception mapper can break the OIDC code flow then it is a flexibility which has become too flexible :-), 

No, this is just a simple case of letting the user handle an advanced use case how they will. If they deliberately try and override the security response and then OIDC no longer works that is on them. Users will always be able to break there app if they do the wrong thing, this is no different to any other feature.

Reply:
@stuartwdouglas 

>If the user wants to handle this we should let them, and get out of their way, trying to 'helpfully' meddle in the process will just break things.

The realistic case I'm having in mind is this one: we have users combining `Basic` and `OIDC web-app`. If the user registers `AuthenticationFailedException` (to customize Basic response error) then two problems appear:
- The user has to figure out which mechanism has thrown the exception
- The user has no way not to pull out of creating OIDC redirect URL

The first one is actually resolvable because `RoutingContext` has an attribute set indicating which `HttpAuthenticationMechanism` initiated the challenge.
The second one is a real one though because it will be hard if not impossible to correctly calculate esp if PKCE is involved.

The same 2nd problem would exist for any combination of interactive and non-interactive mechanisms (basic + form, jwt+webauthn, etc) 

So my proposal is not about interfering into this process but letting users write the exception mappers which should return either null or rethrow the exception - indicating the mapping has not been done and for the default handler take over